components/layout_2020/geom.rs

```diff
@@ -784,25 +792,49 @@ impl Size<LengthPercentage> {
         self.to_numeric()
             .and_then(|length_percentage| length_percentage.to_percentage())
     }
-}
 
-impl LogicalVec2<Size<LengthPercentage>> {
-    pub(crate) fn maybe_percentages_relative_to_basis(
-        &self,
-        basis: &LogicalVec2<Option<Au>>,
-    ) -> LogicalVec2<Size<Au>> {
-        LogicalVec2 {
-            inline: self
-                .inline
-                .maybe_map(|v| v.maybe_to_used_value(basis.inline))
-                .unwrap_or_default(),
-            block: self
-                .block
-                .maybe_map(|v| v.maybe_to_used_value(basis.block))
-                .unwrap_or_default(),
+    /// Resolves percentages in a preferred size, against the provided basis.
+    /// If the basis is missing, percentages are considered cyclic.
+    /// <https://www.w3.org/TR/css-sizing-3/#preferred-size-properties>
+    /// <https://www.w3.org/TR/css-sizing-3/#cyclic-percentage-size>
+    #[inline]
+    pub(crate) fn resolve_percentages_for_preferred(&self, basis: Option<Au>) -> Size<Au> {
+        match self {
+            Size::Numeric(numeric) => numeric
+                .maybe_to_used_value(basis)
+                .map_or(Size::Initial, Size::Numeric),
+            Size::FitContentFunction(numeric) => {
+                // Under discussion in https://github.com/w3c/csswg-drafts/issues/11805
+                numeric
+                    .maybe_to_used_value(basis)
+                    .map_or(Size::FitContent, Size::FitContentFunction)
+            },
+            _ => self.map(|_| unreachable!("This shouldn't be called for keywords")),
         }
     }
 
+    /// Resolves percentages in a maximum size, against the provided basis.
+    /// If the basis is missing, percentages are considered cyclic.
+    /// <https://www.w3.org/TR/css-sizing-3/#preferred-size-properties>
+    /// <https://www.w3.org/TR/css-sizing-3/#cyclic-percentage-size>
+    #[inline]
+    pub(crate) fn resolve_percentages_for_max(&self, basis: Option<Au>) -> Size<Au> {
+        match self {
+            Size::Numeric(numeric) => numeric
+                .maybe_to_used_value(basis)
+                .map_or(Size::Initial, Size::Numeric),
+            Size::FitContentFunction(numeric) => {
+                // Under discussion in https://github.com/w3c/csswg-drafts/issues/11805
+                numeric
+                    .maybe_to_used_value(basis)
+                    .map_or(Size::MaxContent, Size::FitContentFunction)
+            },
+            _ => self.map(|_| unreachable!("This shouldn't be called for keywords")),
```

Comment:
What's the reason for the `self.map`? Could just be `_ => unreachable!(...)` directly?

Reply:
It's not unreachable. `map()` converts e.g. `Size::MinContent` of type `Size<LengthPercentage>` into `Size::MinContent` of type `Size<Au>`. It's just that the mapping function won't be called for keywords.